pyintegration/deephaven2/dtypes.py

```diff
@@ -99,31 +75,226 @@ def array_from(self, values: Iterable):
-single = DType.single
-float32 = DType.float32
-double = DType.double
-float64 = DType.float64
-string = DType.string
-BigDecimal = DType.BigDecimal
-StringSet = DType.StringSet
-DateTime = DType.DateTime
-Period = DType.Period
+class IntegerDType(DType):
+    def __init__(self, qst_type, j_name, np_dtypes: Tuple = (), null_value=None):
+        super().__init__(qst_type=qst_type, j_name=j_name, is_primitive=True)
+        self.np_types = np_dtypes
+        self.null_value = null_value
+
+    def array_of(self, seq: Sequence):
+        """ Creates a Java array of the same integer type populated with values from a Python sequence.
+
+        Args:
+            seq: a Python sequence of compatible data type, can be a np array or Pandas series
+
+        Returns:
+            a Java array
+
+        Raises:
+            DHError
+        """
+        try:
+            if isinstance(seq, np.ndarray):
+                if seq.dtype in self.np_types:
+                    return super().array_of(seq)
+                elif np.issubdtype(seq.dtype, np.integer):
+                    return super().array_of(seq.astype(self.np_types[0]))
+                elif np.issubdtype(seq.dtype, np.floating):
+                    nan_arr = np.isnan(seq)
+                    arr = seq.astype(self.np_types[0])
+                    arr[nan_arr] = self.null_value
+                    return super().array_of(arr)
+                else:
+                    raise ValueError(f"Incompatible np dtype ({seq.dtype}) for {self.j_name} array")
+            elif isinstance(seq, pd.Series):
+                return self.array_of(seq.values)
+            else:
+                return super().array_of(seq)
+        except Exception as e:
+            raise DHError(e, f"failed to create a Java {self.j_name} array.") from e
+
+
+class FloatingDType(DType):
+    def __init__(self, qst_type, j_name, np_dtypes: Tuple = (), null_value=None):
+        super().__init__(qst_type=qst_type, j_name=j_name, is_primitive=True)
+        self.np_types = np_dtypes
+        self.null_value = null_value
+
+    def array_of(self, seq: Sequence):
+        """ Creates a Java array of the same floating type populated with values from a Python sequence.
+
+        Args:
+            seq: a Python sequence of compatible data type, can be a np array or Pandas series
+
+        Returns:
+            a Java array
+
+        Raises:
+            DHError
+        """
+        try:
+            if isinstance(seq, np.ndarray):
+                if seq.dtype in self.np_types:
+                    nan_arr = np.isnan(seq)
+                    arr = seq.copy()
+                    arr[nan_arr] = self.null_value
+                    return super().array_of(arr)
+                elif np.issubdtype(seq.dtype, np.floating):
+                    nan_arr = np.isnan(seq)
+                    arr = seq.astype(self.np_types[0])
+                    arr[nan_arr] = self.null_value
+                    return super().array_of(arr)
```

Comment:
As indicated above, this stuff gives me the willies and needs to be looked at extremely carefully.  See notes above.

Here there are certainly bugs.  For float types, we are setting NaN values to null.  They should remain as NaNs, since the float types support nans.